### ISSUE TITLE
Regression: Out arguments in many COM interfaces (eg. DXGI/D3D11) are suddenly unmanaged

### ISSUE THREAD

**Issue by VentuzTammoHinrichs** [microsoft/CsWin32#1185]
## Actual behavior

COM Out arguments in methods in the Graphics.DXGI and .D3D11 namespaces are now `I..._unmanaged**` (instead of `out I...` in 0.3.49beta). Look at e.g. all the `Create*` functions in `ID3D11Device`.

Note that this doesn't apply to top level functions such as `D3D11CreateDevice`. Those still work as expected.

## Expected behavior

I'd really like to have the managed out arguments back if possible.

## Repro steps

1. `NativeMethods.txt` content:
```
Windows.Win32.Graphics.Direct3D11
Windows.Win32.Graphics.DXGI
D3DCompile
```

2. `NativeMethods.json` content (if present):
```
{
  "$schema": "https://aka.ms/CsWin32.schema.json",
  "useSafeHandles": false,
  "allowMarshaling": true
}
```

### Context

CsWin32 version: 0.3.106
Target Framework: net8.0

**Comment by AArnott** [comment]
I suspect this change is a fallout of improving support in other areas where we discovered certain interop rules had to be honored, and this was an area where we weren't fully following the rules, causing failures in at least some cases.

I'll leave this active till we have a chance to confirm.

**Comment by iraqigeek** [comment]
having the same issue with D3D11 after updating from 0.3.49-beta to 0.3.106

**Comment by smourier** [comment]
Same for me. It's actually a showstopper for upgrading nuget packages.

**Comment by mitchcapper** [comment]
@AffluentOwl if you can update the title to just says something like "regression/breaking change: Multiple prior COM out objects are now _unmanaged double pointers" as it goes beyond the DX items.     This only happened in the latest release 106 the 0.3.49 beta did not.

As this is expected it seems mostly just a few more lines of code, although one does need to track and cleanup the pointer as well now I believe.  Here is a define around the old and new code.

```csharp

            var FilePath=@"c:\users\image.jpg";
            PInvoke.CoCreateInstance<IThumbnailCache>(CLSID_LocalThumbnailCache, null, Windows.Win32.System.Com.CLSCTX.CLSCTX_INPROC_HANDLER | Windows.Win32.System.Com.CLSCTX.CLSCTX_INPROC_SERVER, out var thumCache);
            IShellItem shellItem = GetShellItem(FilePath);
            var size = 256u;
#if NEW_CSWIN32
            ISharedBitmap_unmanaged* img;
            thumCache.GetThumbnail(shellItem, size, WTS_FLAGS.WTS_INCACHEONLY,&img);
            img->GetSharedBitmap(out var hBitmap);           
#else
            thumCache.GetThumbnail(shellItem, size, WTS_FLAGS.WTS_INCACHEONLY,out var img);
            img.GetSharedBitmap(out var hBitmap);
#endif
```

**Comment by VentuzTammoHinrichs** [comment]
It was me. Changed the title.

And I also would like to bump this issue. Yes, you can work with the unmanaged pointers somehow but it's so cumbersome I'd rather write a C++ wrapper DLL now instead of using CsWin32, which kind of runs counter to its point.

**Comment by jimm98y** [comment]
Is there a mode switch that would toggle in between the old and new behavior? Starting with version 0.3.106 there are some functions that have incorrect signature, possibly as a result of this change: https://github.com/microsoft/CsWin32/issues/1229

**Comment by jevansaks** [comment]
I believe this PR is what changed the behavior: https://github.com/microsoft/CsWin32/pull/1109

As an enhancement on top of the PR I've got out (#1511), I can look into making a follow-up PR to improve the friendly overloads of out/ref parameters which are managed types. I'll also think about if there is a way to improve the ABI signature on the COM methods so that implementing these is not so onerous.

**Comment by jimm98y** [comment]
Yes, switching to pointers will be tedious, but when I finally decided to do it so that I can move from the beta to the latest release of CsWin32, I found out that I can't because of that broken MFTEnumEx function. I suppose this will not be the only function with this problem if the metadata are correct.

That was the main reason why I was asking about a way to switch to the old behavior, hoping it would also fix the function import. 

Now I only have 2 options: keep using the 0.3.49 beta version, or write my own imports manually and stop using CsWin32.

**Comment by jevansaks** [comment]
Yeah ...the friendly methods just mis-handle basically any pointer-to-pointer types at the moment.

**Comment by mitchcapper** [comment]
Not quite as set and forget but https://github.com/mono/CppSharp I have had some good luck with and has a higher level of customization.  I am not sure if I am missing something but It would seem the CsWin32 could do overloads with the old style and still provide the pointer option for those desiring to go that route, there isn't a reason it couldn't be an option too as CSWin32 just does on some other things.

**Comment by mitchcapper** [comment]
> Yeah ...the friendly methods just mis-handle basically any pointer-to-pointer types at the moment.

Doesn't doing [In, Out] parameter attributes for the native signature allow for some pointer to pointer types to properly work?

**Comment by jevansaks** [comment]
Yes, but we don't have control over those annotations, they come from win32metadata. 

If you look at the PR I linked above, I have an approach for handling this for optional ref unmanaged-pointer parameter types so after that's merged in I will take a look at extending the logic to handle optional managed-pointer parameters.

**Comment by VentuzTammoHinrichs** [comment]
@jevansaks The original issue isn't fixed yet (tested with 0.3.248) - all ID3D11Device Create* methods still use _unmanaged** arguments for their optional COM outputs.

**Comment by jimm98y** [comment]
You are right, but at least now it generates the correct signature that matches the function definition. And to convert from the _unmanaged** argument I was able to use the same approach mentioned in the comments above:
```cs
ID3D11Texture2D_unmanaged* captureTexture;
_device.CreateTexture2D(&captureTextureDesc, null, &captureTexture);
_captureTexture = (ID3D11Texture2D)Marshal.GetObjectForIUnknown((nint)captureTexture);
```
https://github.com/jimm98y/SharpMediaFoundationInterop/blob/c9087a764504a95f7daa846b41a86559389d04d9/src/SharpMediaFoundationInterop/Input/ScreenCapture.cs#L147

So far it seems to be working.

**Comment by VentuzTammoHinrichs** [comment]
Ok, that looks manageable [pun sorta intended], and I also know that optional out arguments are not a thing in C#. But perhaps it's possible to generate helper functions for this, or have a per-namespace switch in the config that just skips the [Optional] part.

**Comment by VentuzTammoHinrichs** [comment]
> ID3D11Texture2D_unmanaged* captureTexture;
> _device.CreateTexture2D(&captureTextureDesc, null, &captureTexture);
> _captureTexture = (ID3D11Texture2D)Marshal.GetObjectForIUnknown((nint)captureTexture);

Just so you know - it works but you might want to call Release on the unmanaged pointer afterwards; `Marshal.GetObjectForIUnknown` increases the ref count and if I don't release the pointer here, D3D reports a lot of leaks.

**Comment by jimm98y** [comment]
Yes, I am calling Marshal.ReleaseComObject() in my Dispose - that part of code did not have to be changed by updating to the new CsWin32. However, I agree that now the COM type handling is more difficult as now we have both managed and unmanaged variants of the same type, with some functions requiring the managed variant and some functions requiring the unmanaged variant, leaving the marshaling in between them up to the developer.

**Comment by VentuzTammoHinrichs** [comment]
What I was trying to say is that your code example leaks the texture because `Marshal.GetObjectForIUnknown` calls AddRef(). You can't just leave the unmanaged pointer dangling without releasing it, too. (yes, even if you call ReleaseComObject on the managed reference)

**Comment by jimm98y** [comment]
Ah! Thank you!!!

**Comment by jevansaks** [comment]
@jimm98y , @VentuzTammoHinrichs The interface methods have pointers because the COM marshalling support doesn't allow for Optional Ref parameters. But if you enable friendlyOverloads (the default) then you will get extension methods that improve the parameters and do the marshalling for you. Look for this in the generated code:

```c#
internal static unsafe void CreateTexture2D(this winmdroot.Graphics.Direct3D11.ID3D11Device @this, in winmdroot.Graphics.Direct3D11.D3D11_TEXTURE2D_DESC pDesc, winmdroot.Graphics.Direct3D11.D3D11_SUBRESOURCE_DATA? pInitialData, out winmdroot.Graphics.Direct3D11.ID3D11Texture2D ppTexture2D)
```

**Comment by VentuzTammoHinrichs** [comment]
Hey,

My bad - I didn't have `Windows.Win32` in my usings so I completely missed the extension methods.

Sorry for the ruckus, and thank you for the fix <3

**Comment by VentuzTammoHinrichs** [comment]
Ok, sorry again, I still got one: Some of the functions in `ID3D11Device`, eg. `CreateComputeShader()`, only get the extension method without the output but the one with the managed output is missing for some reason. That's an annoyance I can easily work around but perhaps you might have a look into why that is.

**Comment by jevansaks** [comment]
Oh, I've been bitten by that before. I'm not sure why the extension methods are in a different namespace. I'll file a separate issue to look into fixing that.

I'll take a look at CreateComputeShader.

**Comment by jimm98y** [comment]
Thank you very much, thanks to the extension methods I was able to remove some of the uses of the _unmanaged types and simplify the code. However, IMFAttributes.GetItemByIndex (see https://github.com/jimm98y/SharpMediaFoundationInterop/blob/5907cb25e3d9ea5ec8cb28e9d8a46d6173ef13bd/src/SharpMediaFoundationInterop/Transforms/MediaTransformBase.cs#L125) seems to be missing an extension that would accept PROPVARIANT type. I have Windows.Win32 in my usings, yet the only method available is returning PROPVARIANT_unmanaged.

**Comment by jevansaks** [comment]
And is the issue with GetItemByIndex that it's taking params by pointer or that it's got a struct called "_unmanaged"? What would you expect the type of that parameter to be? (In the new AOT com source generator marshalling mode I would like to change these params to "out ComVariant", would that meet your needs?)

**Comment by jimm98y** [comment]
Well, it's now taking PROPVARIANT_unmanaged in 2 overloads, once by ref and then by a pointer. I would expect the overload with "ref" to accept "PROPVARIANT" and the one with a pointer to accept "PROPVARIANT_unmanaged". 
Or, because these are not COM objects, have just a single struct called PROPVARIANT and use it everywhere. Currently, both structs seem identical to me except for the name, so I don't understand why there are 2 instead of just 1.

**Comment by smourier** [comment]
FWIW, please :-) don't "hardcode" translating PROPVARIANT into ComVariant, they are really different things although they look alike. And if you ask me, even translating VARIANT to ComVariant should be optional, as I don't think .NET's ComVariant covers all exotic cases.

**Comment by jevansaks** [comment]
They're ABI compatible so you can reinterpret one to the other. The question is what the default should be in the ABI. ComVariant does handle many PROPVARIANT things (e.g. Dispose does PropVariantClear, so it's PROPVARIANT-aware).

**Comment by smourier** [comment]
Working on Windows, using a VARIANT for a PROPVARIANT and the reverse is at best useless because depending on the API you target, it just won't work (you want VT_LPWSTR, VT_LPSTR, VT_BLOB, VT_VECTOR, etc. support for PROPVARIANT for example with Media Foundation and you'll have VT_BSTR, VT_ARRAY | VT_UI1 etc. support with VARIANT) and at worst a recipe for disaster, just saying.

**Comment by jevansaks** [comment]
> FWIW, please :-) don't "hardcode" translating PROPVARIANT into ComVariant, they are really different things although they look alike. And if you ask me, even translating VARIANT to ComVariant should be optional, as I don't think .NET's ComVariant covers all exotic cases.

@smourier , sure that's reasonable. I'm not familiar with the nuances of PROPVARIANT so I can leave it for now. Maybe we generate some extension methods to convert between the two with like PROPVARIANT.AsComVariant() that returns `ref ComVariant` and ComVariant.AsPROPVARIANT() just to enable easier conversions?

**Comment by smourier** [comment]
Why not, I'm personnally fond of keeping raw native look first (using raw structs that match native structs) and then add extension methods (from CsWin32 or from custom code)

I actually looked at the .NET's `ComVariant` implementation source, and contrary to what I initially thought, it seems it's an hybrid of VARIANT and PROPVARIANT support. However it's not documented as such, so it's difficult to determine what it does support and doesn't w/o looking at its source.

**Comment by jevansaks** [comment]
I believe I've improved this as much as we can for now. Feel free to keep the discussion or report more issues as you encounter them. Thanks!
